### PR TITLE
ADAPT-3250: Adjustments to Tailwind configuration approach to support JIT Mode.

### PR DESCRIPTION
**Pull request by nplowman**
# READY FOR REVIEW

# Summary
- Updates Tailwind inclusion approach. Rather than import fully-compiled CSS from decanter, it now pulls in individual Tailwind plugins and components, as we are doing in the saa_alumni repo.
- Also updates Tailwind to use Just-in-time (JIT) compiler.

# Needed By (Date)
- By end of sprint

# Urgency
- Not aware of any particular urgency. But there may be other tickets waiting on this change as a dependency. 

# Steps to Test

1. Just view Storybook in the preview environment, and compare to https://decanter-react.netlify.app/ to ensure configuration changes did not introduce style regressions.

# Affected Projects or Products
- Yes - any projects using decanter-react will be eventually be impacted by this change after it is merged and projects update to latest version. 

# Associated Issues and/or People
- https://stanfordits.atlassian.net/browse/ADAPT-3250

# Other Notes
- Some npm packages had to be updated to due to dependency requirements for tailwindcss.
- I decided to stay on the previous version of decanter for now. (1.0.0-alpha.7) Let me know if I should upgrade to 1.0.0-beta.0, which is the version of decanter used on saa_alumni.

### CODE REVIEW

File: postcss.config.js
Comment (nplowman) [inline postcss.config.js:4]: Noting that I removed the purgecss configuration, as it was causing build issues when used alongside Tailwind. 
My understanding of why this was being included was because we were including the entire compiled stylesheet from Decanter, which was slowing down hot reloading and increasing the size of build files. So this seems like it would now be irrelevant given this change.

But let me know if there's context I'm missing and I can work out how to add it back in.

Comment (yvonnetangsu) [reply postcss.config.js:4]: @nplowman In our SAA repo, when we first started implementing the JIT mode, it was on a much earlier version of TW when they first introduced the JIT feature. So we found that we still need those settings otherwise the variant classes (the ones with the :) didn't get generated properly. This might have been fixed in the later version of TW.

**Review by yvonnetangsu — CHANGES_REQUESTED** [review]
This looks great to me, @nplowman! Only change request is if we could install the latest version of TW since it comes with some new features that we need. Thanks!

**Comment by yvonnetangsu** [comment]
@rebeccahongsf Tagging you so you're in the loop. Nathan is working on getting JIT mode and this is almost done. You could probably hold off on the button updates in your updating DR to use Decanter beta branch till this get merged. Thanks! 🙏🏼 😄

File: postcss.config.js
Comment (yvonnetangsu) [reply postcss.config.js:4]: Looks like we already got rid of this completely in the SAA repo, so definitely safe to remove. Thanks!

File: postcss.config.js
Comment (sherakama) [inline postcss.config.js:11]: I think we will still want a safelist of colors as I don't think we have them all in our source at the moment and they are a common thing to use as input examples. Is the safelist working in JIT now?

File: tailwind.config.js
Comment (sherakama) [inline tailwind.config.js:12]: Would globbing help with any/all of these require statements? Right now it is a very manual process of adding and removing files as they come and go and this could be a problem for upgrade paths down the line. Perhaps globbing the files might be better?

File: postcss.config.js
Comment (sherakama) [reply postcss.config.js:4]: Yes

File: postcss.config.js
Comment (yvonnetangsu) [reply postcss.config.js:11]: Yes, theoretically safelist should be working in JIT mode now after version 2.2
https://blog.tailwindcss.com/tailwindcss-2-2#built-in-safelist-transform-and-extract-support

Good point about the colors. Perhaps we can safe list the `su-bg-somecolor` and the `su-text-somecolor` classes 🤔

**Comment by yvonnetangsu** [comment]
Hi @nplowman ! Consolidating our feedback/request:

1. Please update to latest version of Tailwind
2. Please see @sherakama's comment about globbing - if we could get all the "require" from Decanter core all at once in tailwind.config.js instead of adding them manually one by one.
3. Please investigate if we can safelist the `su-bg-somecolor` and `su-text-somecolor` classes
https://blog.tailwindcss.com/tailwindcss-2-2#built-in-safelist-transform-and-extract-support

Thank you!

**Comment by nplowman** [comment]
@sherakama @yvonnetangsu 
Those three issues have been resolved.

Note on #3 regarding the safelist, Tailwind's documentation notes that regular expressions do not work in the JIT mode. So instead I'm programmatically crawling through the list of colors to create an explicit safelist for background and text colors.

File: tailwind.config.js
Comment (nplowman) [reply tailwind.config.js]: Globbing and dynamical loading has been added.

File: postcss.config.js
Comment (nplowman) [reply postcss.config.js:11]: Safelist has been moved to tailwind.config.js.

**Comment by yvonnetangsu** [comment]
> @sherakama @yvonnetangsu
> Those three issues have been resolved.
> 
> Note on #3 regarding the safelist, Tailwind's documentation notes that regular expressions do not work in the JIT mode. So instead I'm programmatically crawling through the list of colors to create an explicit safelist for background and text colors.

Thank you for the changes, @nplowman ! I'm out today, but this appears to be working great after playing with it for 5 minutes. I'll do a real review when I come back tomorrow.

**Review by sherakama — APPROVED** [review]
LG2M

File: tailwind.config.js
Comment (yvonnetangsu) [inline tailwind.config.js:56]: @nplowman This all looks good to me. Just a question - why is the logo file treated differently?

**Review by yvonnetangsu — APPROVED** [review]
Pre-approving with just one question. Maybe @sherakama could answer too.

File: tailwind.config.js
Comment (nplowman) [reply tailwind.config.js:56]: Thanks for catching this. I didn't intend to commit this - I was running into errors when it hit the logo plugin, and added this in while I was debugging. I'll remove this.

Comment (yvonnetangsu) [reply tailwind.config.js:56]: Awesome - thank you, @nplowman !

Comment (nplowman) [reply tailwind.config.js]: @yvonnetangsu Debug code has been removed.

**Comment by yvonnetangsu** [comment]
Looking at the logo - I'm remembering that we have 2 more issues that we need to deal with 😬 

1. We need to import the "Stanford" font somehow. Currently, the logo isn't using the custom font (https://deploy-preview-62--decanter-react.netlify.app/?path=/story/stanford-identity-global-footer--default).This is the CSS from Decanter v7:
https://github.com/SU-SWS/decanter/blob/v7/src/css/fonts.css
2. We found that the custom `su-link-somecolor`or  `hover:su-link-underline` utility that we added on Decanter v7 doesn't play nice with the JIT mode. we dealt this this issue in SAA by adding custom styles that overrides the `su-link-somecolor` classes. You could see the issue in the above global footer, where if you hover over one link, all the links are underlined.

**Comment by nplowman** [comment]
@yvonnetangsu 
I can just import the fonts.css directly from Decanter in js file and it should get picked up by the bundler. Where would the best place be to include global css includes like this?
It doesn't look like there are any subdirectories under the src folder for global stuff. Should I just create a subdirectory? 

I have it in .storybook/preview.js at the moment, which works for storybook. But I don't think it will end up in the exported library if it is placed there.

**Comment by yvonnetangsu** [comment]
> @yvonnetangsu
> I can just import the fonts.css directly from Decanter in js file and it should get picked up by the bundler. Where would the best place be to include global css includes like this?
> It doesn't look like there are any subdirectories under the src folder for global stuff. Should I just create a subdirectory?
> 
> I have it in .storybook/preview.js at the moment, which works for storybook. But I don't think it will end up in the exported library if it is placed there.

Hi @nplowman Please feel free to create a subdirectory where you think makes the most sense, and yes, we're going to need it both for Storybook preview and in the exported library. Thanks.

**Comment by nplowman** [comment]
@yvonnetangsu I'm almost there with the changes....
I just spent a couple hours going in circles trying to figure out why the base.js file with the footer link overrides wasn't working  when I tried to use it alongside the glob-imported the plugins from Decanter, even though it worked fine when I tried importing it on its own.

It looks like this plugin from Decanter conflicts with it when it is imported: https://github.com/SU-SWS/decanter/blob/7.0.0-alpha.7/src/plugins/utilities/link/link-underline.js 

On SAA Alumni, this plugin is not imported from Decanter, and is instead replaced with this:
https://github.com/SU-SWS/saa_alumni/blob/main/src/tailwind/plugins/components/underline/link-underline.js

Any thoughts on how we should handle this here?
I could add an exception to the glob import to exclude link-underline.js as is being done in saa_alumni.

But it seems like this maybe should be fixed from within the decanter repo instead.

**Comment by yvonnetangsu** [comment]
> @yvonnetangsu I'm almost there with the changes....
> I just spent a couple hours going in circles trying to figure out why the base.js file with the footer link overrides wasn't working when I tried to use it alongside the glob-imported the plugins from Decanter, even though it worked fine when I tried importing it on its own.
> 
> It looks like this plugin from Decanter conflicts with it when it is imported: https://github.com/SU-SWS/decanter/blob/7.0.0-alpha.7/src/plugins/utilities/link/link-underline.js
> 
> On SAA Alumni, this plugin is not imported from Decanter, and is instead replaced with this:
> https://github.com/SU-SWS/saa_alumni/blob/main/src/tailwind/plugins/components/underline/link-underline.js
> 
> Any thoughts on how we should handle this here?
> I could add an exception to the glob import to exclude link-underline.js as is being done in saa_alumni.
> 
> But it seems like this maybe should be fixed from within the decanter repo instead.

Ah good question. @sherakama , are you ok with removing the custom link-underline.js (and the link-color ones as well) from Decanter v7? Those unfortunately don't play nice with the JIT mode.

**Comment by sherakama** [comment]
Ok, I had a quick look at our codebase. It doesn't appear that we use these classes extensively. I did find a couple of uses in the saa_alumni repo and in the decanter-react repo.

Plan of action:
[ ] Remove link underline file from decanter core
[ ] Add removal notice to UPGRADE.md in decanter core
[ ] Publish a new beta release of decanter core
[ ] Replace all occurrences of `link-underline` and `link-no-underline` in saa_alumni with a working equivalent. 
[ ] Replace all occurrences of `link-underline` and `link-no-underline` in decanter-react with a working equivalent.
[ ] Update this PR with new version of Decanter core and punch a new `patch` release

**Comment by sherakama** [comment]
@nplowman and @yvonnetangsu please have a review.

I was able to get the package to compile with a couple of updates and a few tweaks to the bundler settings. @nplowman I wasn't able to reproduce an error with this package and the link utility in decanter core. Can you provide steps to reproduce?

**Comment by yvonnetangsu** [comment]
Thanks, @sherakama! I'll take a look tomorrow.

**Comment by sherakama** [comment]
Ah, I see the issue with the outputted CSS styles now. 
Screen Shot 2021-09-16 at 10 40 37 PM

It gets a double hover...

**Comment by sherakama** [comment]
@yvonnetangsu please re-review. I think I got to all the parts that were removed in https://github.com/SU-SWS/decanter/pull/851

Merging is blocked until that above PR is merged and a release is cut.

**Comment by sherakama** [comment]
@yvonnetangsu revised and ready for review.

File: src/LocalFooter/LocalFooter.stories.js
Comment (yvonnetangsu) [inline src/LocalFooter/LocalFooter.stories.js:8]:
```suggestion
  '<h2 class="su-font-serif su-text-20 su-tracking-none">Information for</h2><ul class="su-list-unstyled"><li><a class="su-underline hocus:su-no-underline" href="/advisors/">Advisors</a></li><li><a class="su-underline hocus:su-no-underline" href="/alumni/">Alumni</a></li><li><a class="su-underline hocus:su-no-underline" href="/reunion/">Alumni in reunion</a></li><li><a href="https://cfr.stanford.edu/" class="su-underline hocus:su-no-underline su-link--external">Institutional funders</a></li><li><a class="su-underline hocus:su-no-underline" href="/how-to-make-a-gift/international/">International donors</a></li><li><a class="su-underline hocus:su-no-underline" href="/parents-families/">Parents and families</a></li><li><a class="su-underline hocus:su-no-underline" href="/students/">Students</a></li></ul>',
```

File: src/LocalFooter/LocalFooter.stories.js
Comment (yvonnetangsu) [inline src/LocalFooter/LocalFooter.stories.js:9]:
```suggestion
  '<h2 class="su-font-serif su-text-20 su-tracking-none">Information for</h2><ul class="su-list-unstyled"><li><a class="su-underline hocus:su-no-underline" href="/advisors/">Advisors</a></li><li><a class="su-underline hocus:su-no-underline" href="/alumni/">Alumni</a></li><li><a class="su-underline hocus:su-no-underline" href="/reunion/">Alumni in reunion</a></li><li><a href="https://cfr.stanford.edu/" class="su-underline hocus:su-no-underline su-link--external">Institutional funders</a></li><li><a class="su-underline hocus:su-no-underline" href="/how-to-make-a-gift/international/">International donors</a></li><li><a class="su-underline hocus:su-no-underline" href="/parents-families/">Parents and families</a></li><li><a class="su-underline hocus:su-no-underline" href="/students/">Students</a></li></ul>',
```

File: src/LocalFooter/LocalFooter.stories.js
Comment (yvonnetangsu) [inline src/LocalFooter/LocalFooter.stories.js:10]:
```suggestion
  '<h2 class="su-font-serif su-text-20 su-tracking-none">Information for</h2><ul class="su-list-unstyled"><li><a class="su-underline hocus:su-no-underline" href="/advisors/">Advisors</a></li><li><a class="su-underline hocus:su-no-underline" href="/alumni/">Alumni</a></li><li><a class="su-underline hocus:su-no-underline" href="/reunion/">Alumni in reunion</a></li><li><a href="https://cfr.stanford.edu/" class="su-underline hocus:su-no-underline su-link--external">Institutional funders</a></li><li><a class="su-underline hocus:su-no-underline" href="/how-to-make-a-gift/international/">International donors</a></li><li><a class="su-underline hocus:su-no-underline" href="/parents-families/">Parents and families</a></li><li><a class="su-underline hocus:su-no-underline" href="/students/">Students</a></li></ul>',
```

File: src/LocalFooter/LocalFooter.stories.js
Comment (yvonnetangsu) [inline src/LocalFooter/LocalFooter.stories.js:11]: ``
```suggestion
  '<h2 class="su-font-serif su-text-20 su-tracking-none">Information for</h2><ul class="su-list-unstyled"><li><a class="su-underline hocus:su-no-underline" href="/advisors/">Advisors</a></li><li><a class="su-underline hocus:su-no-underline" href="/alumni/">Alumni</a></li><li><a class="su-underline hocus:su-no-underline" href="/reunion/">Alumni in reunion</a></li><li><a href="https://cfr.stanford.edu/" class="su-underline hocus:su-no-underline su-link--external">Institutional funders</a></li><li><a class="su-underline hocus:su-no-underline" href="/how-to-make-a-gift/international/">International donors</a></li><li><a class="su-underline hocus:su-no-underline" href="/parents-families/">Parents and families</a></li><li><a class="su-underline hocus:su-no-underline" href="/students/">Students</a></li></ul>',
```

**Comment by yvonnetangsu** [comment]
Looks great - thank you for all the work, @sherakama and @nplowman! I'll likely do some work to abstract out a link list component to be used in the local footer and we can add the link color change in back that way later on. This is GTG 🚀 !

**Comment by sherakama** [comment]
> I'll likely do some work to abstract out a link list component to be used in the local footer and we can add the link color change in back that way later on

That is a great idea. This PR gets us going but isn't perfect by any stretch.

**Comment by yvonnetangsu** [comment]
> > I'll likely do some work to abstract out a link list component to be used in the local footer and we can add the link color change in back that way later on
> 
> That is a great idea. This PR gets us going but isn't perfect by any stretch.

This PR got us unblocked and should make it much easier for us to build things on. Thank you!

**Comment by yvonnetangsu** [comment]
Please feel free to merge whenever you're ready (just in case any last minute fixup coming in)

**Pull request merged**
